typing/printtyp.ml

```diff
@@ -874,28 +878,59 @@ end = struct
     if non_gen then new_weak_name ty ()
     else new_name ()
 
-  let name_of_type name_generator t =
+  let name_of_type ~non_gen name_generator t =
     (* We've already been through repr at this stage, so t is our representative
        of the union-find class. *)
     let t = substitute t in
     try List.assq t !names with Not_found ->
       try TransientTypeMap.find t !weak_var_map with Not_found ->
       let name =
         match t.desc with
-          Tvar (Some name) | Tunivar (Some name) ->
-            (* Some part of the type we've already printed has assigned another
-             * unification variable to that name. We want to keep the name, so
-             * try adding a number until we find a name that's not taken. *)
+          Tvar (Some name) | Tunivar (Some name) when not non_gen ->
```

Comment:
Reading the code, I don't understand the `not non_gen` business -- this is the only part of the PR I find obscure. Can you expain what is going on?

Reply:
This is to retain a certain behavior that existed before this patch. Basically, we always want to use `name_generator` for non-generalized variables: these are variables representing weak polymorphism instead of proper polymorphism. Without this extra little bit, I found the error messages worse. (I don't remember the details now.)